packages/matter.js/src/device/AutoDevice.ts

```diff
@@ -0,0 +1,361 @@
+
+        // Target prototype for state class
+        public stateProto: any,
+    ) {}
+
+    // Auto-wire state and device interfaces for a specific cluster
+    wire() {
+        this.cluster.attributes.forEach((attr) => {
+            this.wireAttributeState(attr);
+        });
+        this.wireAttributeEvents();
+
+        this.cluster.commands.forEach((command) => {
+            this.wireCommand(command);
+        });
+        
+        this.cluster.events.forEach((event) => {
+            this.wireEvent(event);
+        });
+
+        this.wireClusterServer();
+    }
+
+    // Implement the state interface
+    private wireAttributeState(attr: AttributeModel) {
+        const cluster = this.cluster;
+
+        // Getter
+        const descriptor = <PropertyDescriptor>{
+            get(this: BaseState) {
+                return this[DEVICE].attributeFor(cluster, attr).get();
+            }
+        };
+
+        // Setter
+        if (attr.fixed) {
+            descriptor.set = function(this: BaseState, value) {
+                this[DEVICE].attributeFor(cluster, attr).setLocal(value);
+                return true;
+            }
+        } else {
+            descriptor.writable = false;
+        }
+
+        // Add the property
+        Object.defineProperty(this.stateProto, attr.name, descriptor);
+
+        // If there's a default "get" handler, install it
+        const handler = cluster.defaultServerHandlers[attr.getter];
+        if (handler) this.serverHandlers[attr.getter] = handler;
+    }
+
+    // Implement attribute change events
+    private wireAttributeEvents() {
+        const cluster = this.cluster;
+
+        this.initializers.push(function(this: AutoDevice) {
+            cluster.attributes.forEach((attr) => {
+                const clusterAttr = this.attributeFor(cluster, attr);
+                clusterAttr.addListener((...args) =>
+                    (<any>this)[attr.handler](...args));
+            });
+        });
+    }
+
+    // Implement device commands
+    private wireCommand(command: CommandModel) {
+        this.deviceProto[command.invoker] = ignore;
+        this.serverHandlers[command.handler] =
+            async (request: any, _session, _message, endpoint: any) =>
+                await endpoint[command.handler](request);
+    }
+
+    // Implement device events
+    private wireEvent(event: EventModel) {
+        // TODO - finish once InteractionServer has event support
+        this.deviceProto[event.invoker] = notImplemented;
+    }
+
+    // Wire cluster server installation
+    private wireClusterServer() {
+        const cluster = this.cluster;
+        const handlers = this.serverHandlers;
+        this.initializers.push(function(this: AutoDevice, options: AutoDeviceOptions<BaseState>) {
+            this.addClusterServer(ClusterServer(
+                cluster.definition,
+                Object.assign({}, cluster.defaultState, options.state || {}),
+                handlers));
```

Comment:
Ok now I think I see your approach ...

my current approach was that for many commands we can do a default implementation and it is then very likely that devs do not need to do anything. best example is onWithTimedOff or such on OnOffLighing cluster. and the whole Scenes, Groups and such. So my idea was that we have a defualt implementation that "can check" if there is a better one from the dev, else we use the default. my question here is ... how we could realize this with your approach? Mainly: Where should we put these default and how to wire them? Else we would nee a "getDefualtHandlerFor ClusterrXY" and then the logic here?

Reply:
I agree with your idea.  It would've been clearer if I had finished command wiring but ran out of wifely patience IRL.

If we can leave aside the implementation for a second, here's the shape of the command API that I think would be lowest-friction for the user.  Note that I added one additional feature - the ability for a user handler to fall back to the default handler.

```js
class MyLight extends OnOffLightDevice {
    onOff() {
        if (isNighttimeHours())
            throw new Error("Leave the lights on it's scary out there");
        super();
    }

    // onOn not implemented - just default handler is fine

    onToggle() {
        if (this.state.onOff) {
            // Default implementation doesn't call onOn/onOff
            this.onOff();
            return;
        }
        super();
    }
}
```

I like this because it builds on native language constructs so the syntax is concise and hopefully intuitive.  But it's also powerful in that it gives a lot of flexibility with extension points.

WDYT?  There's a number of ways to implement but probably makes sense to agree if this is the goal first.

...

Now, r.e. implementation...  We could have the device inspect the prototype chain and only add handlers for commands that have been overridden.  That was actually my original plan.  It would make the implementation more complicated (OK for the right API) but the real downside is that invoking super() wouldn't work.

So the solution I came up with was to "proxy" to the real handlers through the device class.  I'm not sure this would work but I looked pretty carefully through the existing handlers and though it might have a chance.

Another idea would be to add a "pre-invoke" handler into CommandServer that could tell the CommandServer whether to invoke the default handler.

Note 1. I had originally taken out scenes and groups from the public API and we could do that again.  I just added back to see what it looks like with them implemented this way.

Note 2. I lean toward making "too many" extension points available rather than too few, so long as they're invisible unless the user wants them.  So I might add interfaces to the device API even if they're completely handled by matter.js.  Can't count the number of times I've had to monkey patch some library because I needed something the author had decided was internal.

Reply:
Another point - the approach I took allows us to support any cluster even if the server isn't implemented for it yet.  If there was a generic `createDefaultClusterServer` it could do the same thing.

It could also add "Not implemented" handlers for commands that don't have a handler.  Then I could take that logic out of "AutoDevice".

Which reminds me, maybe need a better name for "AutoDevice".

Reply:
Oookk, Re concept:
I love that idea in general and for a developer that develops "a special device" this is the most convenient way and I like it.

But when I think that I'm a bridge builder that wants to "bind" several different devicetypes to an other kind of structure (like zwave, zigbee, or myself with iobroker) the idea of inheritence will not help me and so this idea (alone) very much limiting these kind of devs.
The reality right now is that more devs work on/plan bridges is giving this topic a big prio.
For them we would more need a handler approach that I can implement my binding logic once and just "add the handler" to the device class.

With reading the implementation option then they are all not that nice or very complicated or limiting ... Is it that worth "just" for "the most ideal interface?... i'm unsure
A simple "handler approach" (do nmot get me wring, I'm not fighting for it, but it wopuld be a more simple alternative) could still allow devs to implement own methods on the classes and bind them as handlers in the constructor.

Reply:
Also, for completeness:  What do you think about this builder API for bridges?  It would work exactly the same except there would be static methods for installing handlers.

This is really just syntactical sugar for installing handlers.

```js
let device = AutoDevice.implement(chooseDeviceType(zwaveNode));
if (zwaveNode.commandClasses.SwitchBinary) {
    device = device.with(OnOff)
        .onOnOffChange((value) => { /* update z-wave device */ });
    // Map events from zwave -> OnOffInterface
}
if (zwaveDevice.commandClasses.SwitchMultilevel) {
    device = device.with(LevelControl)
        .onLevelChange(/* ... */);
    // ...
}
// Etc.
```

Reply:
Ok a bit abstract. It i think I get the idea. Could work. I think we need that in an example and in the end would throw against @vves in first place  ;-))

Reply:
👍  I think there's 3 API shapes we're considering.  I'll call them "big constructor", "big body" and "builder".  We can default to "big constructor" if the other two don't work out.  They aren't mutually exclusive so we can always start with easiest and get fancier later.